contracts/vaults/RibbonThetaVault.sol

```diff
@@ -321,76 +318,128 @@ contract RibbonThetaVault is DSMath, OptionsVaultStorage {
-    ) private {
-        require(!depositReceipt.processed, "Processed");
-        require(depositReceipt.round < currentRound, "Round not closed");
-        require(depositReceipt.amount > 0, "!amount");
-
-        uint256 pps = roundPricePerShare[depositReceipt.round];
-        // If this throws, it means that vault's roundPricePerShare[currentRound] has not been set yet
-        // which should never happen.
-        // Has to be larger than 1 because `1` is used in `initRoundPricePerShares` to prevent cold writes.
-        require(pps > PLACEHOLDER_UINT, "Invalid pps");
-
-        // This flag avoid overwriting the processed flag when we are creating
-        // a new deposit receipt
-        // saves 300 gas
-        if (updatesDepositReceipt) {
-            depositReceipts[msg.sender].processed = true;
-        }
+        // This handles the null case when depositReceipt.round = 0
+        // Because we start with round = 1 at `initialize`
+        require(depositReceipt.round < round, "Round not closed");
+
+        uint128 unredeemedShares = _getSharesFromReceipt(depositReceipt);
+
+        shares = isMax ? unredeemedShares : shares;
+        require(shares > 0, "!shares");
+        require(shares <= unredeemedShares, "Exceeds available");
 
-        uint256 shares =
-            uint256(depositReceipt.amount).mul(10**uint256(_decimals)).div(pps);
+        depositReceipts[msg.sender] = VaultDeposit.DepositReceipt({
+            processed: true,
+            round: depositReceipt.round,
+            amount: depositReceipt.amount,
+            unredeemedShares: uint128(uint256(unredeemedShares).sub(shares))
+        });
 
         emit Redeem(msg.sender, shares, depositReceipt.round);
 
         _transfer(address(this), msg.sender, shares);
     }
 
+    /**
+     * @notice Returns the shares unredeemed by the user given their DepositReceipt
+     * @param depositReceipt is the user's deposit receipt
+     * @return unredeemedShares is the user's virtual balance of shares that are owed
+     */
+    function _getSharesFromReceipt(
+        VaultDeposit.DepositReceipt memory depositReceipt
+    ) private view returns (uint128 unredeemedShares) {
+        if (!depositReceipt.processed) {
+            uint256 pps = roundPricePerShare[depositReceipt.round];
+
+            // If this throws, it means that vault's roundPricePerShare[currentRound] has not been set yet
+            // which should never happen.
+            // Has to be larger than 1 because `1` is used in `initRoundPricePerShares` to prevent cold writes.
+            require(pps > PLACEHOLDER_UINT, "Invalid pps");
+
+            uint256 sharesFromRound =
+                uint256(depositReceipt.amount).mul(10**uint256(_decimals)).div(
+                    pps
+                );
+            require(sharesFromRound < type(uint104).max, "Overflow");
+
+            uint256 unredeemedShares256 =
+                uint256(depositReceipt.unredeemedShares).add(sharesFromRound);
```

Comment:
is there a possible double counting here? sharesFromRound takes in `depositReceipt.amount` but the deposit amount respective shares could already be reflected in the existing `depositReceipt.unredeemedShares` so we would be adding depositReceipt.unredeemedShares + sharesFromRound and therefore doublecounting?

Reply:
There wasn't an issue with double counting because we check if `depositReceipt.processed` is true. But I made the change to zero out the `depositReceipt.amount` whenever [we do a redemption](https://github.com/ribbon-finance/ribbon-v2/pull/16/commits/7d701ba2c8a4fa9aec93cf84f10f5b410e16a9dd#diff-f882ad74759f8aa2abee31cab887af57224330691be48e0a1e7b1a66f2075cbaR396) so we have extra guarantees